src/lib/AutoControlledComponent.js

```diff
@@ -30,27 +30,33 @@ const getDefaultPropName = (prop) => `default${prop[0].toUpperCase() + prop.slic
 
 /**
  * Return the auto controlled state value for a give prop. The initial value is chosen in this order:
- *  - default props
- *  - then, regular props
+ *  - regular props
+ *  - then, default props
+ *  - then, initial state
  *  - then, `checked` defaults to false
  *  - then, `value` defaults to '' or [] if props.multiple
  *  - else, undefined
  *
- *  @param {object} props A props object
  *  @param {string} propName A prop name
- *  @param {boolean} [includeDefaultProps=false] Whether or not to heed the default prop value
+ *  @param {object} [props] A props object
+ *  @param {object} [state] A state object
+ *  @param {boolean} [includeDefaults=false] Whether or not to heed the default props or initial state
  */
-export const getAutoControlledStateValue = (props, propName, includeDefaultProps = false) => {
-  const defaultPropName = getDefaultPropName(propName)
-  const prop = props[propName]
-  const defaultProp = props[defaultPropName]
-
-  const hasProp = prop !== undefined
-  const hasDefaultProp = defaultProp !== undefined
-
-  // defaultProps & props
-  if (includeDefaultProps && !hasProp && hasDefaultProp) return defaultProp
-  if (hasProp) return prop
+export const getAutoControlledStateValue = (propName, props, state, includeDefaults = false) => {
+  // regular props
+  const propValue = props[propName]
+  if (propValue !== undefined) return propValue
+
+  if (includeDefaults) {
+    // defaultProps
+    const defaultProp = props[getDefaultPropName(propName)]
+    if (defaultProp !== undefined) return defaultProp
+
+    // initial state
+    // check for an object as React initializes state as `null` if it is `undefined`
+    const initialState = ({}).toString.call(state) === '[object Object]' && state[propName]
```

Comment:
I'd maybe do something like:
```js
if (state) {
  const initialState = state[propName]
  if (initialState !== undefined) return initialState
}
```

- `state` will only every be: an object, null, or undefined. I think this is a more straightforward check than the `toString` check.
- `if (initialState) return initialState` wouldn't let `null` through, which should be a valid default value.
- It makes it the same style as the `propValue` and `defaultProp` returns.